storage: add runtime support for batch chunk

hangvane opened the pull request:
Add the runtime support for small chunk mergence.

# basic usage

Add the `--batch-size` arg to command to enable chunk mergence for supported conversion types:

```shell
nydus-image create --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source --batch-size 0x100000

nydus-image create --type estargz-rafs --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source.tar.gz --batch-size 0x100000

nydus-image create --type targz-rafs --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source.tar.gz --batch-size 0x100000

nydus-image create --type tar-rafs --bootstrap ~/bootstrap --blob-dir ~/blobs ~/source.tar --batch-size 0x100000

nydusify convert --source node:latest --target node:latest-nydus-batch-256K --batch-size 0x40000
```

# benchmarks

Tested on 4C4G VM, pulled from private harbor registry in the same Campus Network. Cleaned containerd images and nydus-snapshotter cache before each run. Each value is averaged over five runs.

| repositories| command | tags | prefetch | startup e2e time (s) | % | 
|--|--|--|:--:|--:|--:|
| python | python --version | latest-rafs | ✔ | 2.746 | - |
| python | python --version | latest-rafs (second run) | ✔ | 2.995 | +9.08% |
| python | python --version | latest-rafs-batch-64K | ✔ | 2.202 | -19.80% |
| python | python --version | latest-rafs-batch-128K | ✔ | 2.580 | -6.05% |
| python | python --version | latest-rafs-batch-256K | ✔ | 2.272 | -17.27% |
| python | python --version | latest-rafs-batch-512K | ✔ | 2.157 | -21.44% |
| python | python --version | latest-rafs-batch-1M | ✔ | 1.628 | -40.71% |
| python | python --version | latest-rafs | ✖ | 1.369 | - |
| python | python --version | latest-rafs (second run) | ✖ | 1.174 | -14.25% |
| python | python --version | latest-rafs-batch-64K | ✖ | 1.052 | -23.15% |
| python | python --version | latest-rafs-batch-128K | ✖ | 1.081 | -21.05% |
| python | python --version | latest-rafs-batch-256K | ✖ | 1.075 | -21.48% |
| python | python --version | latest-rafs-batch-512K | ✖ | 1.100 | -19.62% |
| python | python --version | latest-rafs-batch-1M | ✖ | 1.100 | -19.60% |
| python | python --version | alpine3.17-rafs | ✔ | 0.784 | - |
| python | python --version | alpine3.17-rafs (second run) | ✔ | 0.803 | +2.42% |
| python | python --version | alpine3.17-rafs-batch-64K | ✔ | 0.811 | +3.41% |
| python | python --version | alpine3.17-rafs-batch-128K | ✔ | 0.818 | +4.30% |
| python | python --version | alpine3.17-rafs-batch-256K | ✔ | 0.764 | -2.58% |
| python | python --version | alpine3.17-rafs-batch-512K | ✔ | 0.776 | -0.96% |
| python | python --version | alpine3.17-rafs-batch-1M | ✔ | 0.750 | -4.33% |
| python | python --version | alpine3.17-rafs | ✖ | 0.850 | - |
| python | python --version | alpine3.17-rafs (second run) | ✖ | 0.911 | +7.17% |
| python | python --version | alpine3.17-rafs-batch-64K | ✖ | 0.805 | -5.22% |
| python | python --version | alpine3.17-rafs-batch-128K | ✖ | 0.834 | -1.86% |
| python | python --version | alpine3.17-rafs-batch-256K | ✖ | 0.813 | -4.35% |
| python | python --version | alpine3.17-rafs-batch-512K | ✖ | 0.829 | -2.47% |
| python | python --version | alpine3.17-rafs-batch-1M | ✖ | 0.805 | -5.21% |
| node | npm -v | latest-rafs | ✔ | 8.395 | - |
| node | npm -v | latest-rafs (second run) | ✔ | 11.928 | +42.08% |
| node | npm -v | latest-rafs-batch-64K | ✔ | 12.704 | +51.32% |
| node | npm -v | latest-rafs-batch-128K | ✔ | 9.614 | +14.51% |
| node | npm -v | latest-rafs-batch-256K | ✔ | 9.164 | +9.15% |
| node | npm -v | latest-rafs-batch-512K | ✔ | 5.605 | -33.24% |
| node | npm -v | latest-rafs-batch-1M | ✔ | 4.192 | -50.07% |
| node | npm -v | latest-rafs | ✖ | 1.593 | - |
| node | npm -v | latest-rafs (second run) | ✖ | 1.595 | +0.11% |
| node | npm -v | latest-rafs-batch-64K | ✖ | 1.576 | -1.07% |
| node | npm -v | latest-rafs-batch-128K | ✖ | 1.506 | -5.48% |
| node | npm -v | latest-rafs-batch-256K | ✖ | 1.517 | -4.78% |
| node | npm -v | latest-rafs-batch-512K | ✖ | 1.543 | -3.14% |
| node | npm -v | latest-rafs-batch-1M | ✖ | 1.589 | -0.22% |
| node | npm -v | 19-alpine-rafs | ✔ | 2.829 | - |
| node | npm -v | 19-alpine-rafs (second run) | ✔ | 2.667 | -5.72% |
| node | npm -v | 19-alpine-rafs-batch-64K | ✔ | 3.423 | +21.01% |
| node | npm -v | 19-alpine-rafs-batch-128K | ✔ | 3.097 | +9.50% |
| node | npm -v | 19-alpine-rafs-batch-256K | ✔ | 3.249 | +14.85% |
| node | npm -v | 19-alpine-rafs-batch-512K | ✔ | 3.054 | +7.95% |
| node | npm -v | 19-alpine-rafs-batch-1M | ✔ | 2.792 | -1.32% |
| node | npm -v | 19-alpine-rafs | ✖ | 1.372 | - |
| node | npm -v | 19-alpine-rafs (second run) | ✖ | 1.414 | +3.06% |
| node | npm -v | 19-alpine-rafs-batch-64K | ✖ | 1.392 | +1.46% |
| node | npm -v | 19-alpine-rafs-batch-128K | ✖ | 1.396 | +1.73% |
| node | npm -v | 19-alpine-rafs-batch-256K | ✖ | 1.430 | +4.19% |
| node | npm -v | 19-alpine-rafs-batch-512K | ✖ | 1.393 | +1.50% |
| node | npm -v | 19-alpine-rafs-batch-1M | ✖ | 1.411 | +2.86% |


This PR is related to https://github.com/dragonflyoss/image-service/pull/1202, https://github.com/dragonflyoss/image-service/issues/884, https://github.com/dragonflyoss/image-service/issues/885, https://github.com/dragonflyoss/Dragonfly2/issues/1858

hangvane commented:
> Good work! This seems to be valuable in most use cases.
> 
> 1. Can we also give the perf result for `wordpress` image?
> 2. What's the network bandwidth limitation in the tests? It seems that the startup time is slowed down when prefetch all is enabled.
> 3. Can we use 1MB as the default `--batch-size` value?

2. The images are pulled from a private harbor registry in the same Campus Network with 100Mbps bandwidth. The benchmark is just a rough estimation for container startup time beacuse: (1) The startup time difference between the same image `latest-rafs` and `latest-rafs (second run)` shows that it is volatile; (2) The benchmark does not utilize the containers to run workloads, but only checks the versions of python and npm, which may amplify the shortcomings of prefetching, which could occupy network and I/O resources to preload useless data and slow down on-demand requests; ③ The benchmark config follows the default [config settings](https://github.com/dragonflyoss/image-service/blob/d75167992339be64880fff27bfe09a94a9f96396/misc/configs/nydusd-config.json), and the default setting of 8 prefetching threads may be too many for the 4C4G VM and network bandwidth. Anyway, the benchmark is just a rough estimation for presentation. We should precisely evaluate the performance later. Maybe it should be assigned to @guangyuan.

1. Wordpress depends on an external database and does not have a short inspection command to exit the container after execution, like `python --version`. We should precisely evaluate the performance of Wordpress later.

3. It is said that the best practice of the size of read amplification is 128K in datacenter scenarios. Just from a performance point of view, I guess 256K-512K would be good as default value. With good prefetch list assisted at build time, maybe 1M is the best.

imeoer commented:
@hangvane 

Ok, thanks, we can set the default value of `--batch-size` to `128k/256k/512k/1MB` in nydusify to compare the wordpress image with the master branch.

Please also add some tests to passing the codecov test (>=25%).

cc @jiangliu

hangvane commented:
@imeoer Could you please provide me with some guidance on adding tests? I'm finding it difficult to follow the existing tests and add similar ones.

imeoer commented:
@hangvane You may refer to some exists tests in `pub(crate) mod tests`  section.
@adamqqqplay Is it `>=25%` test coverage too strict for PR?

adamqqqplay commented:
> @hangvane You may refer to some exists tests in `pub(crate) mod tests` section. @adamqqqplay Is it `>=25%` test coverage too strict for PR?

@imeoer I think 25% is a pretty loose value, meaning that only a few unit tests need to be added for new code. If the PR is a bugfix, there is usually no need to add unit tests. In [Google's best practice](https://testing.googleblog.com/2020/08/code-coverage-best-practices.html), they require at least 60% coverage.

Of course, if it is really hard to add more unit tests, you can also bypass this check and merge PR.

jiangliu commented on storage/src/cache/mod.rs:
```diff
@@ -410,6 +422,42 @@ impl<'a, 'b> ChunkDecompressState<'a, 'b> {
+
+        self.cache
+            .decompress_chunk_data(input, &mut output, c_size != d_size)?;
+        assert!(output.len() == d_size as usize);
```
Please return error instead of assert, otherwise it may cause DoS attack by injecting bad image data.

jiangliu replied:
This assertion depends data from user input, and we can't trust user here.

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -447,14 +447,34 @@ impl BlobCompressionContextInfo {
             uncompressed_size: round_up_4k(blob_info.uncompressed_size()),
             chunk_info_array: chunk_infos,
             chunk_digest_array: Default::default(),
+            batch_info_array: Default::default(),
```
Could you please help to `impl Default for BlobCompressionContext` so we can simplify the code a bit?

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -447,14 +447,34 @@ impl BlobCompressionContextInfo {
+            let inflate_offset = header.s_ci_zran_offset as usize;
+            let inflate_count = header.s_ci_zran_count as usize;
+            let batch_inflate_size =
+                header.s_ci_zran_count as usize * size_of::<BatchInflateContext>();
```
```
            let batch_inflate_size = inflate_count * size_of::<BatchInflateContext>();
```

hangvane replied:
Changed.

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -656,6 +676,29 @@ impl BlobCompressionContextInfo {
+    /// Get whether chunk at `chunk_index` is batch chunk.
+    /// Some chunks build in batch mode can also be non-batch chunks,
+    /// that they are too big to be put into a batch.
+    pub fn get_chunk_is_batch(&self, chunk_index: u32) -> bool {
```
rename as `is_batch_chunk()`?

hangvane replied:
Changed.

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -1510,7 +1629,46 @@ impl BlobMetaChunkArray {
                 index += 1;
             }
+        } else if last_entry.is_batch() {
+            // assert eatch entry in chunks is Batch chunk.
```
`entch` -> `each`?

hangvane replied:
Changed.

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -1510,7 +1629,46 @@ impl BlobMetaChunkArray {
+                index += 1;
+            }
         } else {
+            // assert eatch entry in chunks is not Batch chunk.
```
ditto

hangvane replied:
Changed.

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -1510,7 +1629,46 @@ impl BlobMetaChunkArray {
+            }
+
+            for entry in &chunk_info_array[index..] {
+                if entry.validate(state).is_err() {
```
`entry.validate()?`?

hangvane committed:
rafs: Terminate the build of a buffered batch chunk if a large chunk is encountered

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: add basic runtime support for batch chunk

1. Add function utils for batch chunk for the help to get batch information.
2. Implement `Default` for `BlobCompressionContext` for simplification.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: add runtime support for batch chunk

Add region calculation and batch chunk decompression capability for nydusd.
Do not support prefetch for batch chunk.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
storage: add runtime prefetch support for batch chunk

Add prefetch range calculation for batch chunk.

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane committed:
smoke: add integration test for batch chunk mode

Signed-off-by: Wenhao Ren <wenhaoren@mail.dlut.edu.cn>

hangvane commented:
/retest

hangvane commented:
/retest

jiangliu commented on storage/src/meta/mod.rs:
```diff
@@ -1468,6 +1606,7 @@ impl BlobMetaChunkArray {
         chunks: &[Arc<dyn BlobChunkInfo>],
         max_size: u64,
     ) -> Result<Vec<Arc<dyn BlobChunkInfo>>> {
+        // `batch_end` is only valid for non-batch chunk.
```
How about renaming "batch_end" to "chunk_end"?

yqleng1987 commented:
/retest

jiangliu committed:
Merge branch 'master' into nydusd
